[Bundle 1/n] Prune g_chainman usage related to ::LookupBlockIndex

dongcarl opened the pull request:
Overall PR: #20158 (tree-wide: De-globalize ChainstateManager)

Note to reviewers:
1. This bundle may _apparently_ introduce usage of `g_chainman` or `::Chain(state|)Active()` globals, but these are resolved later on in the overall PR. [Commits of overall PR](https://github.com/bitcoin/bitcoin/pull/20158/commits)
2. There may be seemingly obvious local references to `ChainstateManager` or other validation objects which are not being used in callers of the current function in question, this is done intentionally to **_keep each commit centered around one function/method_** to ease review and to make the overall change systematic. We don't assume anything about our callers. Rest assured that once we are considering that particular caller in later commits, we will use the obvious local references. [Commits of overall PR](https://github.com/bitcoin/bitcoin/pull/20158/commits)
3. When changing a function/method that has many callers (e.g. `LookupBlockIndex` with 55 callers), it is sometimes easier (and less error-prone) to use a scripted-diff. When doing so, there will be 3 commits in sequence so that every commit compiles like so:
	1. Add `new_function`, make `old_function` a wrapper of `new_function`, divert all calls to `old_function` to `new_function` **in the local module only**
	2. Scripted-diff to divert all calls to `old_function` to `new_function` **in the rest of the codebase**
	3. Remove `old_function`

dongcarl committed:
validation: Move FindForkInGlobalIndex to BlockManager

[META] This commit should be followed up by removing the comments and
       assertions meant only to show that the change is correct.

FindForkInGlobalIndex only acts on BlockManager.

Note to reviewers: Since FindForkInGlobalIndex is always called with
::ChainActive() as its first parameter, it is possible to move
FindForkInGlobalIndex to CChainState and remove this const CChain&
parameter to instead use m_chain. However, it seems like the original
intention was for FindForkInGlobalIndex to work with _any_ chain, not
just the current active chain. Let me know if this should be changed.

dongcarl committed:
validation: Move GetLastCheckpoint to BlockManager

[META] This commit should be followed up by removing the comments and
       assertions meant only to show that the change is correct.

GetLastCheckPoint mainly acts on BlockManager.

dongcarl committed:
validation: Move GetSpendHeight to BlockManager

[META] This commit should be followed up by removing the comments and
       assertions meant only to show that the change is correct.

GetSpendHeight only acts on BlockManager.

dongcarl committed:
validation: Move LoadExternalBlockFile to CChainState

[META] This commit should be followed up by removing the comments and
       assertions meant only to show that the change is correct.

LoadExternalBlockFile mainly acts on CChainState.

dongcarl committed:
validation: Move LookupBlockIndex to BlockManager

[META] This commit should be followed up by a scripted-diff commit which
       fixes calls to LookupBlockIndex tree-wide.
[META] This commit should be followed up by removing the comments and
       assertions meant only to show that the change is correct.

LookupBlockIndex only acts on BlockManager.

dongcarl committed:
scripted-diff: Use BlockManager::LookupBlockIndex

[META] In a previous commit, we moved ::LookupBlockIndex to become a
       member function of BlockManager. This commit is split out from
       that one since it can be expressed nicely as a scripted-diff.

-BEGIN VERIFY SCRIPT-
find_regex='LookupBlockIndex' \
    && git grep -l -E "$find_regex" -- src \
        | grep -v '^src/validation\.\(cpp\|h\)$' \
        | xargs sed -i -E "s@${find_regex}@g_chainman.m_blockman.LookupBlockIndex@g"
-END VERIFY SCRIPT-

dongcarl committed:
validation: Use existing chainstate in ChainstateManager::ProcessNewBlockHeaders

[META] This commit should be followed up by removing the comments and
       assertions meant only to show that the change is correct.

dongcarl committed:
validation: Use accessible chainstate in ChainstateManager::ProcessNewBlock

[META] This commit should be followed up by removing the comments and
       assertions meant only to show that the change is correct.

dongcarl committed:
validation: Make CChainState.m_blockman public

dongcarl committed:
validation: Remove global ::ActivateBestChain

Instead use CChainState::ActivateBestChain, which is what the global one
calls anyway.

dongcarl committed:
validation: Pass in chainstate to ::NotifyHeaderTip

[META] This commit should be followed up by removing the comments and
       assertions meant only to show that the change is correct.

dongcarl committed:
validation: Pass in blockman to ContextualCheckBlockHeader

[META] This commit should be followed up by removing the comments and
       assertions meant only to show that the change is correct.

dongcarl committed:
validation: Pass in chainstate to TestBlockValidity

[META] This commit should be followed up by removing the comments and
       assertions meant only to show that the change is correct.

dongcarl committed:
validation: Remove global LookupBlockIndex

MarcoFalke commented on src/init.cpp:
```diff
@@ -699,7 +699,7 @@ static void ThreadImport(ChainstateManager& chainman, std::vector<fs::path> vImp
                 break; // This error is logged in OpenBlockFile
             LogPrintf("Reindexing block file blk%05u.dat...\n", (unsigned int)nFile);
-            LoadExternalBlockFile(chainparams, file, &pos);
+            ::ChainstateActive().LoadExternalBlockFile(chainparams, file, &pos);
```
I know I've asked the same question already, but I didn't get why this can't use `chainman.ActiveChainstate().Load...`.

This should simplify review because it will touch the same lines of code less often.

Also, it shouldn't make review not harder, because everyone knows that there is only one chainman (whether it is called `g_chainman` or `chainman` shouldn't matter). On top you added a runtime assertion, which validates this for all executed code paths in our tests:

```cpp
     assert(std::addressof(::ChainstateActive()) == std::addressof(ActiveChainstate()));

MarcoFalke replied:
Anyway, no big deal if you want to keep it that way. Will review regardless.

Sjors commented on src/validation.cpp:
```diff
@@ -3392,14 +3391,15 @@ std::vector<unsigned char> GenerateCoinbaseCommitment(CBlock& block, const CBloc
 
 //! Returns last CBlockIndex* that is a checkpoint
-static CBlockIndex* GetLastCheckpoint(const CCheckpointData& data) EXCLUSIVE_LOCKS_REQUIRED(cs_main)
+CBlockIndex* BlockManager::GetLastCheckpoint(const CCheckpointData& data)
```
626b709edfe1bd71e6d485c096ff9a52f698a94c nit: move help text to header file

dongcarl replied:
Will do on next push!

Sjors commented on src/validation.cpp:
```diff
@@ -4671,7 +4667,7 @@ void LoadExternalBlockFile(const CChainParams& chainparams, FILE* fileIn, FlatFi
                 if (hash == chainparams.GetConsensus().hashGenesisBlock) {
                     BlockValidationState state;
-                    if (!ActivateBestChain(state, chainparams, nullptr)) {
+                    if (!::ChainstateActive().ActivateBestChain(state, chainparams, nullptr)) {
```
a375714f24b85b6f9ce7e1765a1137e3b85af009 nit: drop `, nullptr`?

Sjors commented on src/validation.h:
```diff
@@ -654,6 +647,8 @@ class CChainState
 
     /**
+     * Find the best known block, and make it the tip of the block chain
+     *
```
a375714f24b85b6f9ce7e1765a1137e3b85af009  Should this comment be preserved?
```
* May not be called with cs_main held. May not be called in a		
* validationinterface callback.
 ```

dongcarl replied:
Unsure, perhaps @jamesob can help expound on this point a little?

jnewbery replied:
It is preserved, at the bottom of this comment.

Sjors reviewed (approved):
ACK e8dedbb064e90aaa8b2540c7524af54bbdd49d0a

Regarding 98f237506b454816599c58e7f4eb2a65bce32fc2's commit comment:
> However, it seems like the original
intention was for FindForkInGlobalIndex to work with _any_ chain, not
just the current active chain. Let me know if this should be changed.

This was introduced in 2014 by @jtimon in #4796, while refactoring `chainActive.FindFork(locator)` to `findForkInGlobalIndex(chainActive, locator);`. Unless @jamesob sees a use for it for snapshot loading, I think dropping the `chain` parameter should be fine.

jnewbery commented:
Concept ACK!

jonatack commented:
Concept ACK

laanwj commented:
Concept ACK

jnewbery commented:
I've reviewed everything up to 94bbdd7134 _validation: Remove global LookupBlockIndex_ and it looks good so far.

Does it make sense to remove the `assert(std::addressof(thing) == std::addressof(thing));` lines as the last commit of this PR? Is there value in keeping those in place until the series of PRs is finished?

dongcarl commented:
> I've reviewed everything up to [94bbdd7](https://github.com/bitcoin/bitcoin/commit/94bbdd7134588961a63a23a5ac93e6503bee47ec) _validation: Remove global LookupBlockIndex_ and it looks good so far.

Yay! ☺️ 

> Does it make sense to remove the `assert(std::addressof(thing) == std::addressof(thing));` lines as the last commit of this PR? Is there value in keeping those in place until the series of PRs is finished?

I did that in a previous version of this PR, but I changed my mind in this version.

The main reason for this is that as we make further changes in the future (e.g. call `LoadExternalBlockFile` on the local chainman in `ThreadImport` in [this commit](https://github.com/bitcoin/bitcoin/pull/20158/commits/9711fac132695c303918f6b6a4df494126b99854)) we still want to make sure that these assertions hold up for correctness.

Another advantage is that leaving these assertions in prevents new PRs from making the class of mistakes fixed in #20323 from passing tests while the de-globalize chainman bundles are being reviewed and merged.

ryanofsky reviewed (approved):
Code review ACK e8dedbb064e90aaa8b2540c7524af54bbdd49d0a

dongcarl commented on src/validation.h:
```diff
@@ -541,18 +541,18 @@ class CChainState
+    //! Reference to a BlockManager instance which itself is shared across all
+    //! CChainState instances. Keeping a local reference allows us to test more
+    //! easily as opposed to referencing a global.
+    BlockManager& m_blockman GUARDED_BY(::cs_main);
```
True, that doesn't seem necessary!

dongcarl commented:
Push e8dedbb064e90aaa8b2540c7524af54bbdd49d0a -> c948ff186b34693e762fa595ddc174cd88bfd89c
- Addressed https://github.com/bitcoin/bitcoin/pull/20749#discussion_r550179512 and https://github.com/bitcoin/bitcoin/pull/20749#discussion_r556051990

ariard commented on src/validation.cpp:
```diff
@@ -167,17 +166,19 @@ namespace {
 {
     AssertLockHeld(cs_main);
 
+    assert(std::addressof(g_chainman.m_blockman) == std::addressof(*this));
```
I think this `assert` can be removed after 98f2375, as `FindForkInGlobalIndex` is moved to `BlockManager` and thus you can now directly access the new local method `BlockManager::LookupBlockIndex`. Asserting `g_chainmain.m_blockman` equivalence is lapsed ? 

Same for `assert`s in `GetSpendHeight` after d34c069 and in `GetLastCheckpoint` after 866834a

dongcarl replied:
I see your point, but I don't think it hurts anything 😁
All of these will be removed in the end with a scripted-diff anyway

ariard reviewed (commented):
Approach ACK in accordance to #20158

jonatack commented:
Concept ACK

jonatack commented on src/validation.h:
```diff
@@ -438,6 +426,21 @@ class BlockManager
+    CBlockIndex* FindForkInGlobalIndex(const CChain& chain, const CBlockLocator& locator) EXCLUSIVE_LOCKS_REQUIRED(cs_main);
+
+    //! Returns last CBlockIndex* that is a checkpoint
+    CBlockIndex* GetLastCheckpoint(const CCheckpointData& data) EXCLUSIVE_LOCKS_REQUIRED(cs_main);
```
64d27fb0b950bcbd87bc7cd0b4d62f47f305ef88 naming and placement suggestions, feel free to ignore

- as the verb is "look up" (not "lookup", which is a noun), suggest naming the function `LookUpBlockIndex`

- it may be nice to place the function next to its similar twins

```diff
+++ b/src/validation.h
@@ -417,6 +417,7 @@ public:
     CBlockIndex* AddToBlockIndex(const CBlockHeader& block) EXCLUSIVE_LOCKS_REQUIRED(cs_main);
     /** Create a new block index entry for a given block hash */
     CBlockIndex* InsertBlockIndex(const uint256& hash) EXCLUSIVE_LOCKS_REQUIRED(cs_main);
+    CBlockIndex* LookUpBlockIndex(const uint256& hash) EXCLUSIVE_LOCKS_REQUIRED(cs_main);
 
     //! Mark one block file as pruned (modify associated database entries)
     void PruneOneBlockFile(const int fileNumber) EXCLUSIVE_LOCKS_REQUIRED(cs_main);
@@ -431,8 +432,6 @@ public:
         const CChainParams& chainparams,
         CBlockIndex** ppindex) EXCLUSIVE_LOCKS_REQUIRED(cs_main);
 
-    CBlockIndex* LookupBlockIndex(const uint256& hash) EXCLUSIVE_LOCKS_REQUIRED(cs_main);
-
     ~BlockManager() {
         Unload();
     }
```

jonatack reviewed (commented):
Thanks for the care in making the changes clear and easier to review. Started by building and reviewing each of the first three commits 64d27fb / a0954ef / 94bbdd7 after rebasing the pull to master, and running the scripted diff in a0954ef

```
$ test/lint/commit-script-check.sh f91587f..7644282
Running script for: 76442824f66d50a3ceed844349f8d13ce8cf6c87
find_regex='LookupBlockIndex' \
    && git grep -l -E "$find_regex" -- src \
        | grep -v '^src/validation\.\(cpp\|h\)$' \
        | xargs sed -i -E "s@${find_regex}@g_chainman.m_blockman.LookupBlockIndex@g"
OK
```

ryanofsky reviewed (approved):
Code review ACK c948ff186b34693e762fa595ddc174cd88bfd89c. Just lock annotation tweak and added comment since last review

felipsoarez reviewed (commented):
Concept ACK

dongcarl commented:
Note: If #20946 gets in before this, we won't need this last commit I just pushed.

jnewbery commented on src/txmempool.cpp:
```diff
@@ -625,7 +625,7 @@ void CTxMemPool::check(const CCoinsViewCache *pcoins) const
 
     CCoinsViewCache mempoolDuplicate(const_cast<CCoinsViewCache*>(pcoins));
-    const int64_t spendheight = GetSpendHeight(mempoolDuplicate);
+    const int64_t spendheight = WITH_LOCK(::cs_main, return g_chainman.m_blockman.GetSpendHeight(mempoolDuplicate));
```
The lock ordering is to take `cs_main` then take `mempool.cs` so at first glance, this looks like a lock inversion since it's locking `cs_main` while already holding `mempool.cs`. In fact, whenever `mempool.check()` is called, we're already holding `cs_main` (see the two call sites in net_processing).

I think it'd be better to add an `EXCLUSIVE_LOCKS_REQUIRED(cs_main)` annotation to `check()` and remove the re-entrant lock-taking here.

dongcarl replied:
I agree with all the points made, I wanted to keep it this way as it doesn't change the existing (fragile) behaviour, but just makes it more obvious. However, I do think it'd be good to add the annotation.

I opened a new PR for this: https://github.com/bitcoin/bitcoin/pull/20972

dongcarl replied:
Fixed!

jnewbery commented on src/validation.h:
```diff
@@ -50,6 +51,7 @@ struct ChainTxData;
 struct DisconnectedBlockTransactions;
 struct PrecomputedTransactionData;
 struct LockPoints;
+struct CCheckpointData;
```
No need for this forward declaration. You're including chainparams.h, which is where this is included.

dongcarl replied:
Good catch, fixed!

jnewbery commented on src/validation.h:
```diff
@@ -296,7 +289,7 @@ bool UndoReadFromDisk(CBlockUndo& blockundo, const CBlockIndex* pindex);
 
 /** Check a block is completely valid from start to finish (only works on top of our current best block) */
-bool TestBlockValidity(BlockValidationState& state, const CChainParams& chainparams, const CBlock& block, CBlockIndex* pindexPrev, bool fCheckPOW = true, bool fCheckMerkleRoot = true) EXCLUSIVE_LOCKS_REQUIRED(cs_main);
+bool TestBlockValidity(BlockValidationState& state, const CChainParams& chainparams, CChainState& chainstate, const CBlock& block, CBlockIndex* pindexPrev, bool fCheckPOW = true, bool fCheckMerkleRoot = true) EXCLUSIVE_LOCKS_REQUIRED(cs_main);
```
I find these very long lines almost impossible to read. If you touch this branch again, could you be convinced to split it up. Something like:

```suggestion
bool TestBlockValidity(BlockValidationState& state, const CChainParams& chainparams,
                       CChainState& chainstate, const CBlock& block, CBlockIndex* pindexPrev,
                       bool fCheckPOW = true, bool fCheckMerkleRoot = true)
    EXCLUSIVE_LOCKS_REQUIRED(cs_main);
```

dongcarl replied:
Fixed in: d8a5fe24af84c34492e2b23ed341bae70a3c8983

jnewbery commented on src/validation.cpp:
```diff
@@ -3778,42 +3780,44 @@ bool ChainstateManager::ProcessNewBlock(const CChainParams& chainparams, const s
 }
 
-bool TestBlockValidity(BlockValidationState& state, const CChainParams& chainparams, const CBlock& block, CBlockIndex* pindexPrev, bool fCheckPOW, bool fCheckMerkleRoot)
+bool TestBlockValidity(BlockValidationState& state, const CChainParams& chainparams, CChainState& chainstate, const CBlock& block, CBlockIndex* pindexPrev, bool fCheckPOW, bool fCheckMerkleRoot)
```
Same comment as for the declaration. Consider splitting this line.

dongcarl replied:
Fixed in: d8a5fe24af84c34492e2b23ed341bae70a3c8983

jnewbery commented on src/rpc/mining.cpp:
```diff
@@ -375,7 +375,7 @@ static RPCHelpMan generateblock()
 
         BlockValidationState state;
-        if (!TestBlockValidity(state, chainparams, block, LookupBlockIndex(block.hashPrevBlock), false, false)) {
+        if (!TestBlockValidity(state, chainparams, ::ChainstateActive(), block, g_chainman.m_blockman.LookupBlockIndex(block.hashPrevBlock), false, false)) {
```
Any reason not to just jump straight to using `EnsureChainman(request.context)` in all the rpc methods that you're touching here?

dongcarl replied:
I don't think it makes too much of a difference, and I found it clearer to push that decision to a later commit when we've resolved everything under RPC and can see everything that depends on chainman. See: 02e3831e9cd10f78bfde2d67a99c47f08564f07b

jnewbery replied:
ok!

jnewbery reviewed (commented):
I've reviewed everything up to 4124fb0946 _validation: Pass in chainstate to TestBlockValidity_. Just a few additional comments so far.

jnewbery commented on src/validation.h:
```diff
@@ -643,6 +649,8 @@ class CChainState
     void PruneAndFlush();
 
     /**
+     * Find the best known block, and make it the tip of the block chain
```
There's still some redundant text in here. Consider:

```diff
diff --git a/src/validation.h b/src/validation.h
index 10943b6c24..2e1c441ba5 100644
--- a/src/validation.h
+++ b/src/validation.h
@@ -649,18 +649,16 @@ public:
     void PruneAndFlush();
 
     /**
-     * Find the best known block, and make it the tip of the block chain
-     *
-     * Make the best chain active, in multiple steps. The result is either failure
-     * or an activated best chain. pblock is either nullptr or a pointer to a block
-     * that is already loaded (to avoid loading it again from disk).
+     * Find the best known block, and make it the tip of the block chain. The
+     * result is either failure or an activated best chain. pblock is either
+     * nullptr or a pointer to a block that is already loaded (to avoid loading
+     * it again from disk).
      *
      * ActivateBestChain is split into steps (see ActivateBestChainStep) so that
      * we avoid holding cs_main for an extended period of time; the length of this
      * call may be quite long during reindexing or a substantial reorg.
      *
-     * May not be called with cs_main held. May not be called in a
-     * validationinterface callback.
+     * May not be called in a validationinterface callback.
      *
      * @returns true unless a system error occurred
      */
```

dongcarl replied:
Fixed in: d0454bad30253e735254cf7be359212293d952af

jnewbery commented on src/validation.cpp:
```diff
@@ -3650,8 +3651,8 @@ bool ChainstateManager::ProcessNewBlockHeaders(const std::vector<CBlockHeader>&
     }
-    if (NotifyHeaderTip()) {
-        if (::ChainstateActive().IsInitialBlockDownload() && ppindex && *ppindex) {
+    if (NotifyHeaderTip(ActiveChainstate())) {
```
`ActiveChainstate()` doesn't take `cs_main`, whereas `::ChainstateActive()` does. Is that a problem? I imagine we shouldn't be reading `m_active_chainstate` without `cs_main`. The comment says:

```
    //! Once this pointer is set to a corresponding chainstate, it will not
    //! be reset until init.cpp:Shutdown(). This means it is safe to acquire
    //! the contents of this pointer with ::cs_main held, release the lock,
    //! and then use the reference without concern of it being deconstructed.
```

Should `m_active_chainstate` be GUARDED_BY cs_main and `ActiveChainstate()` have EXCLUSIVE_LOCKS_REQUIRED(::cs_main)?

dongcarl replied:
Question: if the concern is that after #19806 the active chainstate may change after “cs_main is released in the code immediately above”, then that concern would still exist even if we did not switch from ::ChainstateActive to ActiveChainstate, no? In other words, this is not a new problem that my PR introduced?

As far as I can tell, the behaviour change that switching from ::ChainstateActive to ActiveChainstate introduces is just a possible race condition in ActiveChainstate() whereby a race condition can occur if m_active_chainstate is modified between its assert and its return, which I’d be happy to fix if we think it’s necessary.

jamesob replied:
For what it's worth, I'd figured that the [`GUARDED_BY` on `g_chainman`](https://github.com/bitcoin/bitcoin/blob/master/src/validation.h#L937) would've cascaded through to many of the usages in question, but maybe I misunderstood.

jnewbery replied:
>  if the concern is that after #19806 the active chainstate may change after “cs_main is released in the code immediately above”, then that concern would still exist even if we did not switch from ::ChainstateActive to ActiveChainstate, no? In other words, this is not a new problem that my PR introduced?

The concern is that reading/writing `m_active_chainstate` is not guaranteed to be atomic. If this thread reads the variable while another thread is writing to it, then it could read the old value, the new value, or anything else. We have no guarantees at all about what is read if the value is not atomic (or guarded by a mutex to prevent multiple threads accessing it at the same time).

This seems fairly simple to solve - just make `m_active_chainstate` atomic.

dongcarl replied:
Fixed by rebasing over #21025, thanks all for the patience and guidance!

jnewbery reviewed (commented):
I've reviewed up to c948ff186b _validation: Use accessible chainstate in ChainstateManager::ProcessNewBlock_.

I'm not very familiar with the fuzzing framework so I didn't review that commit.

ryanofsky commented on src/test/fuzz/load_external_block_file.cpp:
```diff
@@ -15,7 +15,7 @@
 void initialize_load_external_block_file()
 {
-    InitializeFuzzingContext();
+    static TestingSetup setup{CBaseChainParams::REGTEST, {"-nodebuglogfile"}};
```
In commit "fuzz: Initialize a full TestingSetup instead of BasicTestingSetup" (364fa81a2b014bd6257e011e6a201a48c1b8853e)

This seems safe and fine, but I'm only casually familiar with the fuzzing code and can't say I understand what the goal of the change and new code actually is. It might be good to have a code comment saying what the purpose of TestingSetup/REGTEST/nodebuglogfile is here, or a commit description saying what was bad about the previous InitializeFuzzingContext call

dongcarl replied:
Added some context in the commit message here: 4c4eb967016d80247cbf38e6811c8640761ada42
Lmk if that needs additional explanation!

ryanofsky reviewed (approved):
Code review ACK 364fa81a2b014bd6257e011e6a201a48c1b8853e. Only change since last review is new fuzzing commit

laanwj commented:
Code review ACK 364fa81a2b014bd6257e011e6a201a48c1b8853e
Checked that this only moves methods around and doesn't change any validation logic.

kaycoinsUSA commented:
Merge please.

Sent from my iPhone

On Jan 20, 2021, at 12:56 PM, Wladimir J. van der Laan <notifications@github.com> wrote:

﻿

Code review ACK 364fa81<https://na01.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Fbitcoin%2Fbitcoin%2Fcommit%2F364fa81a2b014bd6257e011e6a201a48c1b8853e&data=04%7C01%7C%7Cbb75a1f8661a4fe05e2908d8bd6cbcd2%7C84df9e7fe9f640afb435aaaaaaaaaaaa%7C1%7C0%7C637467622003259859%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwiLCJXVCI6Mn0%3D%7C1000&sdata=XHgPe6fSqzvsnh1XBHfIlLcDFoJ89waX8MwhVHKqYHg%3D&reserved=0>
Checked that this only moves methods around and doesn't change any validation logic.

—
You are receiving this because you are subscribed to this thread.
Reply to this email directly, view it on GitHub<https://na01.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Fbitcoin%2Fbitcoin%2Fpull%2F20749%23issuecomment-763824082&data=04%7C01%7C%7Cbb75a1f8661a4fe05e2908d8bd6cbcd2%7C84df9e7fe9f640afb435aaaaaaaaaaaa%7C1%7C0%7C637467622003269855%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwiLCJXVCI6Mn0%3D%7C1000&sdata=e6NczllcjARVxt57OIGgQXTAuwh1LFahbVEFKQ%2BoEJ4%3D&reserved=0>, or unsubscribe<https://na01.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Fnotifications%2Funsubscribe-auth%2FAPQQKWMMOFWXMSDHY7KC5H3S24KNNANCNFSM4VF7LJHQ&data=04%7C01%7C%7Cbb75a1f8661a4fe05e2908d8bd6cbcd2%7C84df9e7fe9f640afb435aaaaaaaaaaaa%7C1%7C0%7C637467622003269855%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwiLCJXVCI6Mn0%3D%7C1000&sdata=nUXKoVrn0YpE%2BnV9fTaD2jQos%2Fv6dKR99y9QzrjAfnA%3D&reserved=0>.

dongcarl committed:
style-only: Make TestBlockValidity signature readable

dongcarl committed:
style-only: Remove redundant sentence in ActivateBestChain comment

dongcarl commented:
Pushed d0454bad30253e735254cf7be359212293d952af -> 6ea8e92fc1105cf827acf6a72572d16696e2051c
- Rebased over master to include #20946, dropped last commit accordingly

laanwj commented:
re-ACK 6ea8e92fc1105cf827acf6a72572d16696e2051c

glozow reviewed (commented):
Code Review ACK https://github.com/bitcoin/bitcoin/commit/6ea8e92fc1105cf827acf6a72572d16696e2051c

I'm new to the de-globalize chainman effort but fwiw I reviewed each commit noting where things were moving and why they should go there. I can see how `LookupBlockIndex`, `FindForkInGlobalIndex`, `GetSpendHeight`, `GetLastCheckpoint` belong as BlockManager functions. It’s nice to get rid of some globals in validation.cpp, especially with all these very very similar sounding names 🤢 

It makes sense to me to access BlockMap through `g_chainman.blockman` instead of `g_chainman.BlockIndex()` - assertions that they point to the same thing are really helpful - unsure why CChainstateManager and CChainState both have references to the BlockMap in the first place?

dongcarl commented:
Pushed 6ea8e92fc1105cf827acf6a72572d16696e2051c -> 2bda16f4e46bf361bcfa600a4e751e6303a63d0d
- Rebased over master to include #20972, dropped cs_main locking according to https://github.com/bitcoin/bitcoin/pull/20749#discussion_r564711195

glozow commented:
reACK https://github.com/bitcoin/bitcoin/commit/2bda16f4e46bf361bcfa600a4e751e6303a63d0d, only change is dropping repeat-lock of cs_main

laanwj commented:
re-ACK 2bda16f4e46bf361bcfa600a4e751e6303a63d0d

dongcarl committed:
validation: Guard the active_chainstate with cs_main

This avoids a potential race-condition where a thread is reading the
ChainstateManager::m_active_chainstate pointer while another one is
writing to it. There is no portable guarantee that reading/writing the
pointer is thread-safe.

This is also done in way that mimics ::ChainstateActive(), so the
transition from that function to this method is easy.

More discussion:
1. https://github.com/bitcoin/bitcoin/pull/20749#discussion_r559544027
2. https://github.com/bitcoin/bitcoin/pull/19806#discussion_r561023961
3. https://github.com/bitcoin/bitcoin/pull/19806#issuecomment-768946522
4. https://github.com/bitcoin/bitcoin/pull/19806#issuecomment-768955695

dongcarl commented:
If anyone was wondering about resolving the AssumeUTXO changes in #19806 with the deglobalizing-chainman changes, here's an update: https://github.com/bitcoin/bitcoin/pull/19806#issuecomment-768673523.

The conclusion was that the PRs are mergeable in any order without much pain if the last commit of the AssumeUTXO changes in #19806 are slightly modified.

jnewbery commented on src/validation.cpp:
```diff
@@ -3788,42 +3790,50 @@ bool ChainstateManager::ProcessNewBlock(const CChainParams& chainparams, const s
     }
 
-    NotifyHeaderTip();
+    NotifyHeaderTip(ActiveChainstate());
```
Same problem as above: you're grabbing `m_active_chainstate` without locking cs_main first. I think it'd be easiest to make a local `CChainstate*` variable and set it while holding cs_main.

dongcarl replied:
Fixed by rebasing over #21025

dongcarl commented:
Pushed 2bda16f4e46bf361bcfa600a4e751e6303a63d0d -> 246cb0f29b01e9b24cd13388dfd38c4b0729783c
- Rebased on top of #21025 to fix locking
- Addressed jnewbery's review comments (thanks!)

dongcarl commented on src/validation.h:
```diff
@@ -12,6 +12,7 @@
 
 #include <amount.h>
 #include <coins.h>
+#include <chainparams.h>
```
Fixed!

dongcarl commented:
Pushed 246cb0f29b01e9b24cd13388dfd38c4b0729783c -> 67c9a83df19c6e2a2edb32336879204e7770b4a7
- Rebased over new changes in #21025

ryanofsky reviewed (approved):
Code review ACK 67c9a83df19c6e2a2edb32336879204e7770b4a7. Changes since last review:

- new first commit adding guarded by for active chain state, and a few minimal (recursive?) locks to satisfy the annotation. Hopefully these can be replaced later by annotations and appropriately scoped locking
- removing redundant getspendheight lock
- include/comment tweaks

It would be good to see suggestion from https://github.com/bitcoin/bitcoin/pull/20749#discussion_r560979259 followed up at some point: for [`ProcessNewBlock`](
https://github.com/bitcoin/bitcoin/blob/67c9a83df19c6e2a2edb32336879204e7770b4a7/src/validation.cpp#L3794-L3805) or functions like it to avoid calling ActiveChainstate() repeatedly without locking, and instead assign the chainstate they want to use to a variable and use it consistently. (This might be done already in one of the followup PRs)

jnewbery commented:
If @laanwj reACKs this, then I think it'll be ready for merge

laanwj commented:
re-ACK 67c9a83df19c6e2a2edb32336879204e7770b4a7

hebasto commented:
Warnings are fixed in #21051.
